cdap-ui/app/directives/search/search-ctrl.js

```diff
@@ -15,21 +15,120 @@
+      vm.loading = false;
+      vm.searchTerm = '';
+      vm.searchResults = [];
     };
-    this.escapeSearch = function () {
-      if (this.searchTxt === '') {
-        this.displaySearchBox = false;
+
+    vm.search = function (event) {
+      event.stopPropagation();
+      if (event.keyCode === 27) {
+        vm.onBlur();
       }
+
+      if (!vm.searchTerm.length || event.keyCode !== 13) { return; }
+
+      vm.searchEntered = true;
+      vm.loading = true;
+      var params = {
+        namespaceId: $stateParams.namespace || $stateParams.nsadmin,
+        query: vm.searchTerm
+      };
+      myTagsApi.searchTags(params)
+        .$promise
+        .then(function (res) {
+          vm.loading = false;
+          var parsedSearch = [];
+          angular.forEach(res, function (entity) {
+            parsedSearch.push(parseEntity(entity.entityId));
+          });
+
+          vm.searchResults = parsedSearch;
+        });
+
     };
 
+    $scope.$on('$stateChangeSuccess', vm.onBlur);
+
+    // trigger on pressing keyboard '?'
+    $document.bind('keypress', function (event) {
+      if (event.keyCode === 63) {
+        vm.showSearch();
+
+        // for some reason the focus on showSearch() does not trigger the focus
+        $timeout(function () {
+          angular.element(document.getElementById('global-search'))[0].focus();
+        });
+      }
+    });
+
+
+    function parseEntity(entityObj) {
+      switch (entityObj.type) {
+        case 'stream':
+          return {
+            name: entityObj.id.streamName,
+            stateParams: {
+              namespaceId: entityObj.id.namespace.id,
+              streamId: entityObj.id.streamName,
+            },
+            type: 'Stream',
+            icon: 'icon-streams'
+          };
+        case 'datasetinstance':
+          return {
+            name: entityObj.id.instanceId,
+            stateParams: {
+              namespaceId: entityObj.id.namespace.id,
+              datasetId: entityObj.id.instanceId,
+            },
+            type: 'Dataset',
+            icon: 'icon-datasets'
+          };
+        case 'application':
+          return {
+            name: entityObj.id.applicationId,
+            stateParams: {
+              namespaceId: entityObj.id.namespace.id,
+              appId: entityObj.id.applicationId,
+            },
+            type: 'App',
+            icon: 'icon-app'
+          };
+        case 'program':
+          return {
+            name: entityObj.id.id,
+            stateParams: {
+              namespaceId: entityObj.id.application.namespace.id,
+              appId: entityObj.id.application.applicationId,
+              programId: entityObj.id.id,
+            },
+            programType: entityObj.id.type,
+            type: entityObj.id.type,
+            icon: (entityObj.id.type.toLowerCase() === 'flow'? 'icon-tigon' : 'icon-' + entityObj.id.type.toLowerCase())
+          };
+        default:
+          return;
```

Comment:
This will return ```undefined``` for anything other than the above 4 (for instance artifacts). That would cause JS error. we can still return but in the forEach we might want to filter out ```undefined``` as  we can't have ```undefined``` in ```ng-repeat```